### PR TITLE
[Incremental] Enforce thread-safety better for `IncrementalCompilationState` and `ModuleDependencyGraph`

### PR DESCRIPTION
**Pull request by davidungar**
Pave the way for string interning by restructuring the incremental compilation state to clarify the concurrency restrictions. 
The next might be changing the `ModuleDependencyGraph` from a class to a struct to better reflect its mutation.
Then adding a string intern table to the graph, or to the `IncrementalCompilationState.ProtectedState`.
@artemcm , What do you think?

This is a large change; I have tried to structure it into digestible commits.

### CODE REVIEW

**Comment by davidungar** [comment]
@swift-ci please test

**Comment by davidungar** [comment]
@swift-ci please test macos platform

**Comment by davidungar** [comment]
@swift-ci please test macos platform

**Comment by davidungar** [comment]
@swift-ci please test

**Review by artemcm — APPROVED** [review]
This change looks good to me for the most part (some nits and a suggestion below).
This is a nice improvement for reasoning about and enforcing thread-safety of the mutable state that we have here!

**Comment by davidungar** [comment]
Measured performance to ensure no slowdown: 
testCleanBuildSwiftDepsPerformance: This PR was slightly faster (not significantly) or the same.
Building a large 1000+ source file app: 
- This PR was faster by about 20% according to reported compilation time in build log.
- Measuring driver CPU time via Instruments, no difference.

Could be that the extra parallelism here helps.

File: Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift [hunk 1]
```diff
@@ -76,517 +59,47 @@ public final class IncrementalCompilationState {
...
+    
+    reporter?.reportOnIncrementalImports(
+      initialState.incrementalOptions.contains(.enableCrossModuleIncrementalBuild))
 
     let firstWave =
       try FirstWaveComputer(initialState: initialState, jobsInPhases: jobsInPhases,
                             driver: driver, reporter: reporter).compute(batchJobFormer: &driver)
 
-    self.skippedCompileGroups = firstWave.skippedCompileGroups
+    self.protectedState = ProtectedState(
```
[the remark this answers is not in the document]
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState.swift:74]: Good idea! Will do.

File: Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState+Extensions.swift [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+import TSCBasic
+import TSCUtility
+import Foundation
+import SwiftOptions
+
+/// In a separate file to ensure that ``IncrementalCompilationState/protectedState``
+/// can only be accessed via ``IncrementalCompilationState/safelyDo(_:)`` and
```
[the remark this answers is not in the document]
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationState+Extensions.swift:18]: Great catch!

File: Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationProtectedState.swift [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+import TSCBasic
+import TSCUtility
+import Foundation
+import SwiftOptions
+
+/// An instance of `IncrementalCompilationState.Actor` encapsulates the data necessary
+/// to make incremental build scheduling decisions and protects it from concurrent access.
+
+extension IncrementalCompilationState {
+  // MARK: - IncrementalCompilationStateActor
```
[the remark this answers is not in the document]
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationProtectedState.swift:21]: Great catch!

File: Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationProtectedState.swift [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+import TSCBasic
+import TSCUtility
+import Foundation
+import SwiftOptions
+
+/// An instance of `IncrementalCompilationState.Actor` encapsulates the data necessary
```
[the remark this answers is not in the document]
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/IncrementalCompilationProtectedState.swift:17]: Great catch!

**Comment by davidungar** [comment]
> This change looks good to me for the most part (some nits and a suggestion below).
> This is a nice improvement for reasoning about and enforcing thread-safety of the mutable state that we have here!

Thank you for looking at this! Great suggestions!

**Comment by davidungar** [comment]
@swift-ci please test

**Comment by davidungar** [comment]
@swift-ci please test macos platform

**Comment by davidungar** [comment]
@swift-ci please test

**Pull request merged**